module/execution/registers_async.go

```diff
@@ -0,0 +1,66 @@
+package execution
+
+import (
+	"fmt"
+	"sync"
+
+	"go.uber.org/atomic"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/state_synchronization"
+	"github.com/onflow/flow-go/storage"
+)
+
+// RegistersAsyncStore has the same basic structure as access/backend.ScriptExecutor
+// TODO: use this implementation in the `scripts.ScriptExecutor` passed into the AccessAPI
+type RegistersAsyncStore struct {
+	registerIndex storage.RegisterIndex
+	reporter      state_synchronization.IndexReporter
+	initialized   *atomic.Bool
+	init          sync.Once
+}
+
+func NewRegistersAsyncStore() *RegistersAsyncStore {
+	return &RegistersAsyncStore{
+		initialized: atomic.NewBool(false),
+	}
+}
+
+// InitDataAvailable follows the same pattern of backend.ScriptExecutor
+// This method can be called at any time after the RegistersAsyncStore object is created and
+// calls to GetRegisterValues will return a storage.ErrHeightNotIndexed,
+// since we can't disambiguate between the underlying store before bootstrapping or just simply being behind sync
+func (r *RegistersAsyncStore) InitDataAvailable(
+	indexReporter state_synchronization.IndexReporter,
+	registers storage.RegisterIndex,
+) {
+	r.init.Do(func() {
```

Comment:
Why do we need this init bool? Is it to prevent the caller calling multiple times? In that case, why would the caller call multiple times?

Reply:
it's to avoid having to use a lock to protect the `reporter` and `registerIndex`. this way, if `initialized` is true, then these values are set and will never be updated

Reply:
I don't think we need to protect reporter or register index. If `initialized` is false, no API will hit register index or reporter

Reply:
but after it is set once, `initialized` is true so if it is ever set again, there could be a race condition.

If we update this to remove `reporter`, then we could remove `initialized` and make `indexReporter` an atomic pointer. then we can get rid of this

Reply:
Why do we set it again?

Does the caller know it has set it before?

Could we return error if the caller is trying to set twice?

Reply:
it should only ever be set once, that's why this is here. We could definitely switch this to a `CompareAndSwap` on `initialized` or `indexReporter` if it's converted to an atomic pointer, then return an error if it's already set

Reply:
We are going to drop `indexReporter`, and only keep the `registerIndex`.

So we could remove both `init` and `initialized`, and replace the `registerIndex` with a atomic Pointer. If compareAndSwap fails during `InitDataAvailable`, it means it's called multiple times, and should return error. Right?

Reply:
yes exactly. sorry for the confusion

Reply:
it's expected to be called just once after the `RegisterIndex` bootstrap is complete, so the underlying store doesn't get reset after